### ISSUE TITLE
Issue with Creating Transforms 

### ISSUE THREAD

**Issue by JesseNordin** [enotspe/fortinet-2-elasticsearch#54]
We are seeing an issue with the ELK Stack and Creating the Transforms,

Using the Put Command, I have tried to load the first Transform via Dev console and it yields the following: 

{
  "error": {
    "root_cause": [
      {
        "type": "validation_exception",
        "reason": "Validation Failed: 1: Failed to test query, received status: BAD_REQUEST;"
      }
    ],
    "type": "validation_exception",
    "reason": "Validation Failed: 1: Failed to test query, received status: BAD_REQUEST;",
    "caused_by": {
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query",
      "grouped": true,
      "failed_shards": [
        {
          "shard": 0,
          "index": ".ds-logs-fortinet.fortigate.traffic-default-2023.11.10-000034",
          "node": "tGJaL1oXRjepu80fEqDmsQ",
          "reason": {
            "type": "illegal_argument_exception",
            "reason": "Fielddata is disabled on [fgt.srchwvendor] in [.ds-logs-fortinet.fortigate.traffic-default-2023.11.10-000034]. Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [fgt.srchwvendor] in order to load field data by uninverting the inverted index. Note that this can use significant memory."
          }
        }
      ],
      "caused_by": {
        "type": "illegal_argument_exception",
        "reason": "Fielddata is disabled on [fgt.srchwvendor] in [.ds-logs-fortinet.fortigate.traffic-default-2023.11.10-000034]. Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [fgt.srchwvendor] in order to load field data by uninverting the inverted index. Note that this can use significant memory.",
        "caused_by": {
          "type": "illegal_argument_exception",
          "reason": "Fielddata is disabled on [fgt.srchwvendor] in [.ds-logs-fortinet.fortigate.traffic-default-2023.11.10-000034]. Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [fgt.srchwvendor] in order to load field data by uninverting the inverted index. Note that this can use significant memory."
        }
      }
    }
  },
  "status": 400
}

**Comment by JesseNordin** [comment]
JSON File: 

`{
  "id": "endpoint.metadata_united-default-8.7.1",
  "authorization": {
    "roles": [
      "kibana_system"
    ]
  },
  "version": "8.7.1",
  "create_time": 1683815936499,
  "source": {
    "index": [
      "metrics-endpoint.metadata_current_default*",
      ".fleet-agents*"
    ],
    "query": {
      "match_all": {}
    }
  },
  "dest": {
    "index": ".metrics-endpoint.metadata_united_default"
  },
  "frequency": "1s",
  "sync": {
    "time": {
      "field": "updated_at",
      "delay": "4s"
    }
  },
  "pivot": {
    "group_by": {
      "agent.id": {
        "terms": {
          "field": "agent.id"
        }
      }
    },
    "aggregations": {
      "united": {
        "scripted_metric": {
          "init_script": "state.docs = []",
          "map_script": "state.docs.add(new HashMap(params['_source']))",
          "combine_script": "return state.docs",
          "reduce_script": "def ret = new HashMap(); for (s in states) { for (d in s) { if (d.containsKey('Endpoint')) { ret.endpoint = d } else { ret.agent = d } }} return ret"
        }
      }
    }
  },
  "description": "Merges latest Endpoint and Agent metadata documents",
  "settings": {},
  "_meta": {
    "managed_by": "fleet",
    "managed": true,
    "package": {
      "name": "endpoint"
    }
  }
}`

**Comment by JesseNordin** [comment]
When trying to look at the preview I get the following: 

An error occurred loading the index data. Please ensure your query syntax is valid.

Forbidden: [[security_exception] Cannot preview transform [transform-preview] because user elastic lacks the required permissions [.fleet-agents*:[read, view_index_metadata], metrics-endpoint.metadata_current_default*:[]]]: Cannot preview transform [transform-preview] because user elastic lacks the required permissions [.fleet-agents*:[read, view_index_metadata], metrics-endpoint.metadata_current_default*:[]]

**Comment by enotspe** [comment]
Looks like on your `ds-logs-fortinet.fortigate.traffic-default-2023.11.10-000034` index,  `fgt.srchwvendor` field has been created as `text` mapping, not `keyword`.

Make sure you are using [strings_as_keyword@mappings.json](https://github.com/enotspe/fortinet-2-elasticsearch/blob/master/index%20templates/component%20templates/strings_as_keyword%40mappings.json) as part of your [index template](https://github.com/enotspe/fortinet-2-elasticsearch/blob/master/index%20templates/index%20templates/logs-fortinet.fortigate.traffic#L27).

**Comment by JesseNordin** [comment]
Hello @enotspe , Thank you very much. I have managed to add strings_as_keyword@mappings.json as a component of the Index Template and that worked. 

It has prompted me to revise all the Index Templates to look for missing components as we are getting strange errors in the dashboards. 

So far I have found quite few few component templates missing from their respective index templates. I have tried to add them again via the GUI  but I am getting weird errors for it. 

One's missing thus far are: 

"ecs-vulnerability",
"auto_expand_replicas@settings",
"refresh_interval@settings"
"ecs-log-modified",
"ecs-event-modified",
"ecs-file-modified",
"synthetic_source@mappings"
"ecs-network-transforms",
"ecs-source-transforms",
"ecs-event-transforms",
"transforms-fortinet.fortigate@mappings",
"ecs-rule-transforms",

**Comment by JesseNordin** [comment]
I have managed to re-add all component templates and add the relevant component templates to each index template. I did this by simply re running the Put Index template commands and saw no error about missing components. The only thing I have left out is the Auto_Expand Replica component as I am running it on a single node and replica's make the index state yellow. 

When I now look at the transforms dashboards such as Sessions Traffic [Fortigate] [Transforms] I get error's in the data fields saying that the field is missing e.g. Field @timestamp was not found.

I made sure that I do have the transform loded in te ingest pipelines and they are all there. 
E.g. [transforms-fortinet.fortigate.traffic.private2private]

Not quite sure what the issue may be now, do these transforms require replica's ?
